### PR TITLE
Scan backend speedup

### PR DESCRIPTION
**Pull request by carriepl**

### CODE REVIEW

**Comment by carriepl** [comment]
Some tests are failing in FAST_COMPILE. Currently looking at it.

**Comment by carriepl** [comment]
Done and Travis passes.

File: theano/scan_module/scan_perform.pyx [hunk 1]
```diff
@@ -256,9 +254,11 @@ def perform(
...
     other_args = args[offset:]
     input_storage = fnct.input_storage
-    len_input_storage = len(input_storage)
-    old_input_storage = [None] * len_input_storage
-    old_input_data = [None] * len_input_storage
+    nb_mitmot_in = 0
+    for idx in range(n_mit_mot):
+        nb_mitmot_in += tap_array_len[idx]
+    old_mitmot_input_storage = [None] * nb_mitmot_in
+    old_mitmot_input_data = [None] * nb_mitmot_in
```
Comment (abergeron) [inline theano/scan_module/scan_perform.pyx]: You might be able to go even faster by making these C arrays instead of python lists.

**Comment by nouiz** [comment]
I merge now as this PR strictly bring Theano in a better state and it solve a regression in some cases.

The comments are future enhancement.

**Pull request merged**